File: leo2-node/src/test/java/org/deku/leo2/node/messaging/sync/EntitySyncTest.java
package org.deku.leo2.node.messaging.sync;

import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import org.deku.leo2.messaging.activemq.ActiveMQContext;
import org.deku.leo2.node.DataTest;
import org.deku.leo2.node.data.PersistenceContext;
import org.deku.leo2.node.data.entities.Route;
import org.deku.leo2.node.data.sync.EntityConsumer;
import org.deku.leo2.node.data.sync.EntityPublisher;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.slf4j.LoggerFactory;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import javax.persistence.EntityManagerFactory;
import javax.persistence.PersistenceUnit;
import java.net.URI;

/**
 * Created by masc on 18.06.15.
 */
//@Ignore
@RunWith(SpringJUnit4ClassRunner.class)
public class EntitySyncTest extends DataTest {
    @PersistenceUnit(name = PersistenceContext.DB_EMBEDDED)
    private EntityManagerFactory mEntityManagerFactory;

    private EntityPublisher mListener;
    private EntityConsumer mClient;

    @Before
    public void setup() throws Exception {
        Logger l = (Logger) LoggerFactory.getLogger("org.deku.leo2.node");
        l.setLevel(Level.DEBUG);

        // Enforcing tcp connection
        ActiveMQContext.instance().setLocalUri(new URI("tcp://localhost:61616"));
        // Starting broker
        ActiveMQContext.instance().getBroker().start();

        mListener = new EntityPublisher(ActiveMQContext.instance(), mEntityManagerFactory);
        mClient = new EntityConsumer(ActiveMQContext.instance(), mEntityManagerFactory);
    }

    @After
    public void tearDown() {
        mClient.dispose();
        mListener.dispose();
        ActiveMQContext.instance().getBroker().dispose();
    }

    @Test
    public void testSync() throws Exception {
        mListener.start();
        mClient.request(Route.class);

        Thread.sleep(10000);
    }
}

File: leoz-node/src/prototype/java/org/deku/leoz/node/sync/EntitySyncTest.kt
package org.deku.leoz.node.sync

import org.deku.leoz.config.ArtemisConfiguration
import org.deku.leoz.config.JmsEndpoints
import org.deku.leoz.node.config.DataTestConfiguration
import org.deku.leoz.node.config.MessageBrokerTestConfiguration
import org.deku.leoz.node.config.PersistenceConfiguration
import org.deku.leoz.node.data.jpa.MstRoute
import org.deku.leoz.node.service.internal.sync.EntityConsumer
import org.deku.leoz.node.service.internal.sync.EntityPublisher
import org.deku.leoz.node.service.internal.sync.ConsumerPreset
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.experimental.categories.Category
import org.junit.runner.RunWith
import org.slf4j.LoggerFactory
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.junit4.SpringRunner
import sx.mq.MqBroker
import sx.mq.jms.activemq.ActiveMQBroker
import sx.mq.jms.artemis.ArtemisBroker
import sx.junit.PrototypeTest
import sx.mq.jms.JmsEndpoint
import java.util.concurrent.Executors
import javax.inject.Inject
import javax.persistence.EntityManagerFactory
import javax.persistence.PersistenceUnit

/**
 * Created by masc on 18.06.15.
 */
//@Ignore
@Category(PrototypeTest::class)
@RunWith(SpringRunner::class)
@SpringBootTest(classes = arrayOf(
        DataTestConfiguration::class,
        MessageBrokerTestConfiguration::class))
class EntitySyncTest {
    private val log = LoggerFactory.getLogger(this.javaClass)

    @PersistenceUnit(name = PersistenceConfiguration.QUALIFIER)
    private lateinit var entityManagerFactory: EntityManagerFactory

    private var listener: EntityPublisher? = null
    private var client: EntityConsumer? = null

    @Inject
    private lateinit var broker: MqBroker

    @Before
    @Throws(Exception::class)
    fun setup() {
        // Enforcing tcp connection
        //ActiveMQContext.instance().getBroker().setLocalUri(new URI("tcp://localhost:61616"));

        // Starting broker
        this.broker.start()

        val notificationChannelConfig: JmsEndpoint
        val requestChannelConfig: JmsEndpoint
        when (this.broker) {
            is ActiveMQBroker -> {
                notificationChannelConfig = JmsEndpoints.central.entitySync.topic
                requestChannelConfig = JmsEndpoints.central.entitySync.queue
            }
            is ArtemisBroker -> {
                notificationChannelConfig = ArtemisConfiguration.entitySyncTopic
                requestChannelConfig = ArtemisConfiguration.entitySyncQueue
            }
            else -> throw UnsupportedOperationException("Unknown broker type")
        }

        listener = EntityPublisher(
                notificationEndpoint = notificationChannelConfig,
                requestEndpoint = requestChannelConfig,
                entityManagerFactory = entityManagerFactory,
                listenerExecutor = Executors.newSingleThreadExecutor())

        client = EntityConsumer(
                notificationEndpoint = notificationChannelConfig,
                requestEndpoint = requestChannelConfig,
                entityManagerFactory = entityManagerFactory,
                listenerExecutor = Executors.newSingleThreadExecutor(),
                presets = listOf(
                        ConsumerPreset(type = MstRoute::class.java)
                ))
    }

    @After
    fun tearDown() {
        client!!.close()
        listener!!.close()
    }

    @Test
    @Throws(Exception::class)
    fun testSync() {
        listener!!.start()
        for (i in 0..8)
            client!!.request(MstRoute::class.java, null, true)

        Thread.sleep(60000)
    }
}
